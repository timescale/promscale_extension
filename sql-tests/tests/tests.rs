use insta::assert_snapshot;
use std::{
    env,
    mem::MaybeUninit,
    sync::{Mutex, Once},
};
use test_common::*;
use test_generator::test_resources;

const TESTDATA: &'static str = concat!(env!("CARGO_MANIFEST_DIR"), "/testdata");

fn init_lock() -> &'static Mutex<()> {
    static mut SINGLETON: MaybeUninit<Mutex<()>> = MaybeUninit::uninit();
    static ONCE: Once = Once::new();

    unsafe {
        ONCE.call_once(|| {
            let singleton = Mutex::new(());
            SINGLETON.write(singleton);
        });

        SINGLETON.assume_init_ref()
    }
}

// TODO fix how test_resources works in nexted workspaces
#[test_resources("sql-tests/testdata/*.sql")]
fn sql_tests(full_resource: &str) {
    let pg_blueprint = PostgresContainerBlueprint::new().with_testdata(TESTDATA);
    let test_pg_instance = new_test_instance_from_env(&pg_blueprint);
    let mut init_conn = test_pg_instance.connect();
    // This lock prevents multiple callers from running
    // CREATE EXTENSION simultaneously. In case of
    // promscale_extension concurrent CREATE EXTENSION causes
    // issues when migration incremental/003-users.sql executes.
    {
        let _init_lock = init_lock().lock().unwrap();
        init_conn
            .simple_query("CREATE EXTENSION promscale;")
            .expect("Unable to create extension promscale.");
    }

    let resource = if let Some((_, rest)) = full_resource.split_once('/') {
        rest
    } else {
        full_resource
    };
    let query_result = test_pg_instance.exec_sql_script(resource);
    assert_snapshot!(resource, query_result);
}
