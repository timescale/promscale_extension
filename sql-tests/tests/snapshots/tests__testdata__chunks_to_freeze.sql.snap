---
source: sql-tests/tests/tests.rs
expression: query_result
---
 plan 
------
 1..5
(1 row)

   create_hypertable   
-----------------------
 (13,public,metric1,t)
(1 row)

 add_compression_policy 
------------------------
                   1005
(1 row)

   create_hypertable   
-----------------------
 (15,public,metric2,t)
(1 row)

 add_compression_policy 
------------------------
                   1006
(1 row)

         isnt_empty         
----------------------------
 ok 1 - chunks were created
(1 row)

                   is_empty                    
-----------------------------------------------
 ok 2 - zero results when no chunks compressed
(1 row)

                      results_eq                       
-------------------------------------------------------
 ok 3 - 15 uncompressed chunks should have been chosen
(1 row)

                      results_eq                      
------------------------------------------------------
 ok 4 - 15 compressed chunks should have been created
(1 row)

                        results_eq                         
-----------------------------------------------------------
 ok 5 - view should return the 15 chosen compressed chunks
(1 row)

 finish 
--------
(0 rows)


